Add Object::to_iter for retrieving an object's keys and values without allocating a rust collection

daboross opened the pull request:
This could be a bit overzealous, but here it is.

This allows converting an `Object` into a container like `HashMap<i32, V>` rather than only `HashMap<String, V>`, as well as any other map-like thing. It uses a user-provided parsing function to parse the string keys into the other type.

I can make this less permissive if needed - especially since I was only originally thinking about HashMaps.

Fixes #359.

koute commented:
Sorry for the delay and thanks for the PR!

Wow, this is indeed very generic and somewhat complex, and I probably don't really want to expose a `serde`-specific error type here. Hmm... perhaps it would be better to just make a method which simply returns an iterator? E.g. something like this: (`to_iter` and not `iter` because creating such iterator is not zero cost)

```
    fn to_iter( &self ) -> impl ExactSizeIterator< Item = (String, Value) > { ... }
```

then the user would just be able to efficiently convert the keys, convert the values, and collect the collection into whatever they want (a `HashMap`, a `BTreeMap`, or something else).

Such an iterator could be adapted from the existing private `ObjectDeserializer` - basically, 1) convert the slices to raw pointers, and 2) add a `Drop` which will consume the whole iterator (to prevent leaks) and deallocate the arrays for keys and values. That should, *I think*, work. What would you say to this?

daboross commented:
This is great!

I didn't really feel satisfied with forcing reliance on `ConversionError` - and the iterator feels like a cleaner solution overall too.

Since `deserialize_object` has the comment `// TODO: Panic-safety.`, I'm assuming it could benefit from an iterator with a drop implementation too. Would it be alright if I modified `ObjectDeserializer` to implement `Drop` like you described, and then exposed a method like `serialization::deserialize_object` which just returns an `ObjectDeserializer`?

With your first comment - all delays are fine with me! Nothing I'm doing with this is mission-critical, or anything - and I can't claim to be consistently punctual myself.

daboross commented:
> Since `deserialize_object` has the comment `// TODO: Panic-safety.`, I'm assuming it could benefit from an iterator with a drop implementation too. Would it be alright if I modified `ObjectDeserializer` to implement `Drop` like you described, and then exposed a method like `serialization::deserialize_object` which just returns an `ObjectDeserializer`?

I implemented this. Let me know if you'd rather have a separate type and not have it affect the existing deserialization code?

I'm not 100% sure whether panicking in WASM will unwind the stack, but if it does, this change should make `deserialize_object` panic-safe.

daboross commented on src/webcore/serialization.rs:
```diff
@@ -214,13 +215,54 @@ impl SerializedUntaggedArray {
+
+pub struct ObjectDeserializer {
+    key_slice: OwnedFfiSlice < SerializedUntaggedString >,
+    value_slice: OwnedFfiSlice < SerializedValue< 'static > >,
```
I couldn't find any reason why using `'static` here would cause a safety risk, but it seems like it could be wrong in some way since the lifetime parameter does exist there. Since this makes `ObjectDeserializer` own its contents and drop it on drop, it didn't feel right to use any external lifetime parameter here.

koute commented:
The `'static` lifetime looks fine to me in this case.

And I don't think unwinding works in WASM yet (at least it didn't the last time I checked), so on panic the code might leak anyway, but there isn't really anything we can do about it now.

Anyway, looks good to me! Thanks!

daboross commented:
> And I don't think unwinding works in WA SM yet (at least it didn't the last time I checked), so on panic the code might leak anyway, but there isn't really anything we can do about it now.

Makes sense. It'll work once we get real unwinding, at least.

In any case, I'm glad to have this - and with a nicer interface than my original pr! Thanks!
